cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -0,0 +1,337 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clusterstate
+
+import (
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"k8s.io/contrib/cluster-autoscaler/cloudprovider"
+	apiv1 "k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/plugin/pkg/scheduler/schedulercache"
+
+	"github.com/golang/glog"
+)
+
+const (
+	// Maximum time from the moment the node is registered to the time the node is ready.
+	MaxNodeStartupTime = 3 * time.Minute
+)
+
+// ScaleUpRequest contains information about the requested node group scale up.
+type ScaleUpRequest struct {
+	// NodeGroupName is the node group to be scaled up.
+	NodeGroupName string
+	// Time is the time when the request was submitted.
+	Time time.Time
+	// ExpectedAddTime is the time at which the request should be fulfilled.
+	ExpectedAddTime time.Time
+	// NodeTemplate is the template of the node that will appear due to this request.
+	NodeTemplate *schedulercache.NodeInfo
+	// How much the node group is increased.
+	Increase int
+}
+
+// NodeDeletionRequiest contains information about the requested node deletion.
+type NodeDeletionRequest struct {
```

Comment:
The naming is inconsistent: ScaleUpRequest / NodeDeletionRequest. It would be better to have ScaleUpRequest / ScaleDownRequest or NodeAdditionRequest / NodeDeletionRequest.